controllers/external.go

```diff
@@ -0,0 +1,191 @@
+package controllers
+
+import (
+	"context"
+	"errors"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/bakito/vault-unsealer/pkg/cache"
+	"github.com/bakito/vault-unsealer/pkg/constants"
+	"github.com/bakito/vault-unsealer/pkg/types"
+	"github.com/hashicorp/vault-client-go"
+	"golang.org/x/sync/errgroup"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// ExternalHandler handles external vaults
+type ExternalHandler struct {
+	client.Client
+	Scheme     *runtime.Scheme
+	startedMux sync.Mutex
+	started    bool
+	secrets    []corev1.Secret
+	Cache      cache.Cache
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *ExternalHandler) SetupWithManager(mgr ctrl.Manager, secretsExternal []corev1.Secret) error {
+	r.secrets = secretsExternal
+	return mgr.Add(r)
+}
+
+func (r *ExternalHandler) Start(ctx context.Context) error {
+	r.startedMux.Lock()
+	if r.started {
+		return errors.New("handler is already running")
+	}
+	r.started = true
+	r.startedMux.Unlock()
+
+	grp, ctx := errgroup.WithContext(ctx)
+
+	for _, s := range r.secrets {
+		grp.Go(func() error {
+			err := r.setupVaultCheckLoop(ctx, s)
+			if err != nil {
+				return err
+			}
+			return context.Canceled
+		})
+	}
+
+	_ = grp.Wait()
+	return nil
+}
+
+func (r *ExternalHandler) setupVaultCheckLoop(ctx context.Context, secret corev1.Secret) error {
+	if r.Cache.VaultInfoFor(secret.Name) == nil {
+		v := &types.VaultInfo{
+			Username:   string(secret.Data[constants.KeyUsername]),
+			Password:   string(secret.Data[constants.KeyPassword]),
+			Role:       string(secret.Data[constants.KeyRole]),
+			SecretPath: string(secret.Data[constants.KeySecretPath]),
+		}
+
+		for key, val := range secret.Data {
+			if strings.HasPrefix(key, constants.KeyPrefixUnsealKey) {
+				v.UnsealKeys = append(v.UnsealKeys, string(val))
+			}
+		}
+
+		r.Cache.SetVaultInfoFor(secret.Name, v)
+	}
+
+	duration := r.getInterval(secret)
+
+	srcCl, err := r.getSourceClient(secret)
+	if err != nil {
+		return err
+	}
+
+	trgtsCl, err := r.getTargetClients(secret)
+	if err != nil {
+		return err
+	}
+
+	t := time.NewTicker(duration).C
+	for {
+		r.executeCheck(ctx, secret.Name, srcCl, trgtsCl)
+		select {
+		case <-t:
+			continue
```

Comment:
should r.executeCheck not be done within the switch on a new tick?

Reply:
it's acutally works like it is but i agree it's strange. i changed it

Reply:
btw i figured out the difference. the original version executes at start as well while this verison only excutes on the first tick. i will change it so it also executes at startup

Reply:
ctx.Cancel is not needed any more?

Reply:
might be better for the shutdown. changed it again, pls have a look